students/Cracolici_Jon/lesson05/JonCracolici_Mailroom3.py

```diff
@@ -0,0 +1,180 @@
+    dB.update(updict)
+    return dB
+def update_donor(newdondict, dB):
+    """Updates an existing donor in the database with a new donation."""
+    dB[newdondict['donornamekey']]['totaldonated'] += newdondict['donation']
+    dB[newdondict['donornamekey']]['numdonations']+=1
+    dB[newdondict['donornamekey']]['avgdonated'] = dB[newdondict['donornamekey']]['totaldonated']/dB[newdondict['donornamekey']]['numdonations']
+    dB[newdondict['donornamekey']]['lastdonation'] = newdondict['donation']
+    return dB
+def show_current_names(dB):
+    """Displays the names of all previous donors."""
+    names = [item['donor'] for item in dB.values()]
+    #for item in dB.values():
+    #    names.append(item['donor'])
+    lastname = names.pop(len(names)-1)
+    l = len(names)
+    display = "The current donors are"+(l*" {},").format(*names)+" and {}.".format(lastname)
+    print(display)
+    return #Returns you to the send a thank you menu
+    #return display #, saty_control(dB)
+
+def create_report(dB):
+    """Creates a summary report for all current donors."""
+    dictlist = OrderedDict(sorted(dB.items(), key=lambda x: x[1]['totaldonated'], reverse=True))
+    print('Donor Name                | Total Given | Num Gifts | Average Gift')
+    print('------------------------------------------------------------------')
+    for i in dictlist:
+        print('{donor:26}${totaldonated:>13.2f}{numdonations:>12} ${avgdonated:>12.2f}'.format(**dB[i]))
+    return
+def mass_mail(dB):
+    for item in dB.values():
+        print_letters(item)
+    return start_control(dB)
+def nd_control(dB):
+    """Function that calls functions to create and add a new donor to database, print letter to screen."""
+    newdonordict = donor_input(dB)
+    if newdonordict == False:
+        return False
+    elif newdonordict['donornamekey'] in dB.keys():
+        print("It appears that donor is already in our system.\n Perhaps you would like to select the 'Update Donor' option instead?")
+        return False
+    #will add try except to check for donor actually being new in L5
+    new_donor(newdonordict, dB)
+    note_gen(dB[newdonordict['donornamekey']])
+    return dB#, start_control(dB)
+def ud_control(dB):
+    """Function that calls functions to update a donor in the database, print letter to screen."""
+    updonordict = donor_input(dB)
+    if updonordict == False:
+        return False
+    elif updonordict['donornamekey'] not in dB.keys():
+        print("It appears that you are attempting t update a donor not already in our system.\n Perhaps you would like to select the 'New Donor' option instead?")
+        return False
+    #will add try except to check for donor actually being new in L5
+    update_donor(updonordict, dB)
+    note_gen(dB[updonordict['donornamekey']])
+    return dB#, start_control(dB)
+def donor_input(dB):
+    """This function collects user input about the donation."""
+    #create a set of quit criteria
+    quitset = set(['quit', 'q'])
```

Comment:
no need for `set` here; a tuple is more suitable because it's immutable and good for constant type definitions.

Reply:
ok. Just tried it and works fine. What is set good for then? being able to keep checking as things get added?

Reply:
Great question; there are many uses for sets but I will give you an example.

Let's say in another use case we are pulling donations from the database and databases records are returned as lists of lists like:
```
donors = [["Bob", 300], ["Mary", 200], ["Bob", 400]]
```
and you need to get a list of donors (unique)

you could:
```
names = []
for donor in donors:
    if name not in names:
        names.append(donor[0])
```
or
```
names = set()
for donor in donors:
    names.add(donor[0])
```

set code is much simpler and you're utilizing a data structure built for this

Reply:
so sets can be really useful when you are checking things that might have multiple identical entries. Like in that fruit lab exercise.

Reply:
Exactly